ios/ServicexSdkRn.swift

```diff
@@ -79,48 +100,87 @@ class ServicexSdkRn: NSObject {
                 }catch let error as NSError {
                     reject("CredifySDK error","parsing json error", error)
                 }
-            case .failure(let error):
-                print(error)
-                reject("CredifySDK error", "get offerlist error", error)
-            }
-        }
+            }).store(in: &cancellables)
+        
     }
     
     @objc(clearCache)
     func clearCache(){
-        CredifyServiceX.session.resetSession()
+        CredifyServiceX.resetSession()
     }
     
     @objc(setPushClaimRequestStatus:)
     func setPushClaimRequestStatus(isSuccess: Bool){
-        self.pushClaimResponseCallback?(isSuccess)
+        if isSuccess {
+            self.pushClaimResponseCallback?(.success(()))
+        }else {
+            print("push claim token issue")
+            self.pushClaimResponseCallback?(.failure(UnmanagedError.init()))
+        }
         // Dereference callback to avoid memory leak
         self.pushClaimResponseCallback = nil
     }
     
     @objc(showOfferDetail:pushClaimCB:)
     func showOfferDetail(offerId: String, pushClaimCB:@escaping(RCTResponseSenderBlock)) {
-        let user = self.parseUserProfile(value: userInput!)
+        guard let ui = userInput else {
+            print("User input was not found")
+            return
+        }
+        guard let user = self.parseUserProfile(value: ui) else {
+            print("User was not found")
+            return
+        }
+        
         let offer = self.listOffer.first(where: { item -> Bool in
             item.id == offerId
         })
+        
+        guard let offer = offer else {
+            print("Offer was not found")
+            return
+        }
+        
+        guard let vc = UIApplication.shared.keyWindow?.rootViewController else {
+            print("There is no view controller")
+            return
+        }
+        
+        let task = { (credifyId: String) -> Future<Void, Error> in
+            return Future() { promise in
+                pushClaimCB([user.id, credifyId])
+                self.pushClaimResponseCallback = promise
+            }
+        }
+        
         DispatchQueue.main.async {
-            CredifyServiceX.offer.presentModally(from: UIApplication.shared.keyWindow!.rootViewController!, offer: offer!, userProfile: user!){ credifyId, result in
-                // Demo Market call push claim token
-                pushClaimCB([user?.id, credifyId])
-                self.pushClaimResponseCallback = result
-            } completionHandler: { result in
-                print("**** Result is ****")
-                print(result)
+            CredifyServiceX.offer.presentModally(
+                from: vc,
+                offer: offer,
+                userProfile: user,
+                pushClaimTokensTask: task
+            ) { [weak self] result in
+                
```

Comment:
Is there no callback? iOS and Android should receive a callback from integrating services. So RN should also have a callback. This callback will be called when the SDK is dismissed with a result (`success`, `canceled`, `failed`)

Reply:
As I discussed with you before, the callback is the limitation of react-native bridge method. 
In this link here: https://reactnative.dev/docs/0.63/native-modules-ios
It said `A native module is supposed to invoke its callback only once` 

We use the callback for push claim token already so for the callback for the result of the redemption, we need to make another bridge function to do an interval check for the redemption result. I created a ticket here https://app.shortcut.com/credify/story/20330/react-native-sdk-return-transaction-status-after-finishing-redemption-flow and I plan to do it in the next sprint. For now, we test and give this SDK to Med247 first so they can integrate to their mobile app first. In the meanwhile, we will update the code and let them know later. Is it okay ?

Reply:
Your plan to deliver the SDK first is good. I agreed with that. Regarding the work around, the updated iOS logic doesn't use 2 callbacks it seems, we should test it more.